app/models/application_qualification.rb

```diff
@@ -141,4 +141,30 @@ def composite_equivalency_details
 
     details.strip if details.present?
   end
+
+  def set_public_id(preferred_id_value: nil)
+    top_level_public_id = public_id.presence || preferred_id_value.presence || next_available_public_id
+
+    set_public_ids_for_constituent_grades(top_level_public_id)
+    self.public_id = top_level_public_id
+  end
+
+private
+
+  def set_public_ids_for_constituent_grades(top_level_public_id)
+    return if constituent_grades.blank?
+
+    grades_with_ids = constituent_grades.transform_values.with_index do |grade, index|
+      next grade if grade['public_id'].present?
+
+      public_id = index.zero? ? top_level_public_id : next_available_public_id
```

Comment:
This is a bit unclear and prone to issues we've not thought of.

It's here because the method is trying to do two things:
- Account for the initial migration, where we need to copy across the db ids
- Generate new ids for new qualifications

I'm going to split up the above